sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java

```diff
@@ -337,6 +338,18 @@ public Instant apply(KV<byte[], Long> input) {
     }
   }
 
+  // Kafka records are read in a separate thread inside the reader. As a result advance() might not
+  // read any records even from the mock consumer, especially for the first record.
+  // This is a helper method to loop until we read a record.
+  private static void advanceOnce(UnboundedReader<?> reader) throws IOException {
+    int attempts = 0;
+    while (!reader.advance()) {
+      attempts++;
+      // very rarely will there be more than one attempts.
+      assertTrue("could not advance() even after 1000 attempts.", attempts < 1000);
```

Comment:
Seems like this scenario could still happen if the reader thread gets swapped out but the work thread does not.

Rather just delete this check and let test-level timeouts handle the failure case. These have a reasonable default (20s?) and maybe override it below.